pkg/cluster/manager.go

```diff
@@ -0,0 +1,1886 @@
+		return perrs.Trace(err)
+	}
+
+	if err := os.RemoveAll(d.specManager.Path(clusterName, "patch")); err != nil {
+		return perrs.Trace(err)
+	}
+
+	log.Infof("Upgraded cluster `%s` successfully", clusterName)
+
+	return nil
+}
+
+// Patch the cluster.
+func (d *Manager) Patch(clusterName string, packagePath string, opt operator.Options, overwrite bool) error {
+	metadata, err := d.meta(clusterName)
+	if err != nil {
+		return perrs.AddStack(err)
+	}
+
+	topo := metadata.GetTopology()
+	base := metadata.GetBaseMeta()
+
+	if exist := utils.IsExist(packagePath); !exist {
+		return perrs.New("specified package not exists")
+	}
+
+	insts, err := instancesToPatch(topo, opt)
+	if err != nil {
+		return err
+	}
+	if err := checkPackage(d.specManager, clusterName, insts[0].ComponentName(), insts[0].OS(), insts[0].Arch(), packagePath); err != nil {
+		return err
+	}
+
+	var replacePackageTasks []task.Task
+	for _, inst := range insts {
+		deployDir := clusterutil.Abs(base.User, inst.DeployDir())
+		tb := task.NewBuilder()
+		tb.BackupComponent(inst.ComponentName(), base.Version, inst.GetHost(), deployDir).
+			InstallPackage(packagePath, inst.GetHost(), deployDir)
+		replacePackageTasks = append(replacePackageTasks, tb.Build())
+	}
+
+	t := task.NewBuilder().
+		SSHKeySet(
+			d.specManager.Path(clusterName, "ssh", "id_rsa"),
+			d.specManager.Path(clusterName, "ssh", "id_rsa.pub")).
+		ClusterSSH(topo, base.User, opt.SSHTimeout).
+		Parallel(replacePackageTasks...).
+		Func("UpgradeCluster", func(ctx *task.Context) error {
+			return operator.Upgrade(ctx, topo, opt)
+		}).
+		Build()
+
+	if err := t.Execute(task.NewContext()); err != nil {
+		if errorx.Cast(err) != nil {
+			// FIXME: Map possible task errors and give suggestions.
+			return err
+		}
+		return perrs.Trace(err)
+	}
+
+	if overwrite {
+		if err := overwritePatch(d.specManager, clusterName, insts[0].ComponentName(), packagePath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ScaleOutOptions contains the options for scale out.
+type ScaleOutOptions struct {
+	User         string // username to login to the SSH server
+	IdentityFile string // path to the private key file
+	UsePassword  bool   // use password instead of identity file for ssh connection
+}
+
+// DeployOptions contains the options for scale out.
+// TODO: merge ScaleOutOptions, should check config too when scale out.
+type DeployOptions struct {
+	User              string // username to login to the SSH server
+	IdentityFile      string // path to the private key file
+	UsePassword       bool   // use password instead of identity file for ssh connection
+	IgnoreConfigCheck bool   // ignore config check result
+}
+
+// Deploy a cluster.
+func (d *Manager) Deploy(
+	clusterName string,
+	clusterVersion string,
+	topoFile string,
+	opt DeployOptions,
+	afterDeploy func(b *task.Builder, newPart spec.Topology),
+	skipConfirm bool,
+	optTimeout int64,
+	sshTimeout int64,
+	nativeSSH bool,
```

Comment:
All the last four options are contained in `operator.Options`, why not pass the `options` directly?

Reply:
`skipConfirm` is not.
the good thing for keeping the left 3 params is making it clear what this function needs since `operator.Options` contains many unrelate params.

I prefer to just keep it now.